tools/ddprobe.cpp

```diff
@@ -69,9 +69,131 @@ syncThreadDesktop() {
+  const uint8_t *pixels = static_cast<const uint8_t *>(mappedResource.pData);
+  const int bytesPerPixel = 4;  // Assuming RGBA format
+  const int stride = mappedResource.RowPitch;
+  const int width = frameDesc.Width;
+  const int height = frameDesc.Height;
+
+  // Convert the threshold to an integer value for comparison
+  const int threshold = static_cast<int>(blackThreshold * 255);
+
+  // Loop through every pixel
+  for (int y = 0; y < height; ++y) {
+    for (int x = 0; x < width; ++x) {
+      const uint8_t *pixel = pixels + y * stride + x * bytesPerPixel;
+      // Check if any channel (R, G, B) is significantly above black
+      if (pixel[0] > threshold || pixel[1] > threshold || pixel[2] > threshold) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+/**
+ * @brief Attempts to capture and verify the contents of up to 10 consecutive frames from a DXGI output duplication.
+ *
+ * This function tries to acquire the next frame from a provided DXGI output duplication object (`dup`) and inspects its content to determine if the frame is not completely black. If a non-black frame is found within the 10 attempts, the function returns `S_OK`. If all 10 frames are black, it returns `S_FALSE`, indicating that the capture might not be functioning properly. In case of any failure during the process, the appropriate `HRESULT` error code is returned.
+ *
+ * @param dup A reference to the DXGI output duplication object (`dxgi::dup_t&`) used to acquire frames.
+ * @param device A pointer to the ID3D11Device interface that represents the device associated with the Direct3D context.
+ * @return HRESULT Returns `S_OK` if a non-black frame is captured successfully, `S_FALSE` if all frames are black, or an error code if a failure occurs during the process.
+ *
+ * Possible return values:
+ * - `S_OK`: A non-black frame was captured, indicating capture was successful.
+ * - `S_FALSE`: All 10 frames were black, indicating potential capture issues.
+ * - `E_FAIL`, `DXGI_ERROR_*`, or other DirectX HRESULT error codes in case of specific failures during frame capture, texture creation, or resource mapping.
+ */
+HRESULT
+test_frame_capture(dxgi::dup_t &dup, ID3D11Device *device) {
+  for (int i = 0; i < 10; ++i) {
+    std::cout << "Attempting to acquire the next frame (" << i + 1 << "/10)..." << std::endl;
+    IDXGIResource *frameResource = nullptr;
+    DXGI_OUTDUPL_FRAME_INFO frameInfo;
+    HRESULT status = dup->AcquireNextFrame(500, &frameInfo, &frameResource);
+    if (FAILED(status)) {
+      std::cout << "Failed to acquire next frame [0x" << std::hex << status << "]" << std::endl;
+      return status;
+    }
+
+    std::cout << "Frame acquired successfully." << std::endl;
+
+    ID3D11Texture2D *frameTexture = nullptr;
+    status = frameResource->QueryInterface(__uuidof(ID3D11Texture2D), (void **) &frameTexture);
+    frameResource->Release();
```

Comment:
Unsafe resource cleanup. "RAII" based approach should be used, like `util::fail_guard` or `boost::scope::scope_exit` (if available).

For example:
```
IDXGIResource *frame_resource = nullptr;
auto frame_resource_guard{util::fail_guard([&]() {
    if (frame_resource) {
        frame_resource ->Release();
        frame_resource  = nullptr;
    }
}};

Reply:
I've though about ComPTR instead as well, but not surprised I missed that one because I don't really code in c++ often